domain/credential/service/service.go

```diff
@@ -145,6 +149,34 @@ func (s *Service) CloudCredentialsForOwner(ctx context.Context, owner user.Name,
 	return result, nil
 }
 
+// SingleValidCloudCredentialsForOwner returns the owner's valid (and non-revoked) credential key and cloud credentials for a given cloud
+// if only one credential exists. Where owner has multiple valid cloud credentials an error is returned.
+func (s *Service) SingleValidCloudCredentialsForOwner(ctx context.Context, owner user.Name, cloudName string) (corecredential.Key, cloud.Credential, error) {
+	ctx, span := trace.Start(ctx, trace.NameFromFunc())
+	defer span.End()
+
+	returnError := func(err error) (corecredential.Key, cloud.Credential, error) {
+		return corecredential.Key{}, cloud.Credential{}, err
+	}
+
+	creds, err := s.st.ValidCloudCredentialsForOwner(ctx, owner, cloudName)
+	if err != nil {
+		return returnError(errors.Capture(err))
+	}
+	if len(creds) > 1 {
+		return returnError(credentialerrors.MultipleCloudCredentialsFound)
+	}
+	for name, credInfoResult := range creds {
+		key, err := corecredential.KeyFromString(name)
+		if err != nil {
+			return returnError(errors.Capture(err))
+		}
+		return key, cloudCredentialFromCredentialResult(credInfoResult), nil
+	}
+
+	return returnError(credentialerrors.NotFound)
```

Comment:
We don't return raw error types with no context. This will be extremely hard to trace down in the future.